backend/src/main/java/ca/mcgill/purposeful/service/ModeratorService.java

```diff
@@ -0,0 +1,168 @@
+ *
+ * @author Enzo Benoit-Jeannin
+ */
+@Service
+public class ModeratorService {
+
+  // CRUD Repositories
+  @Autowired
+  private AppUserRepository appUserRepository;
+
+  @Autowired
+  private ModeratorRepository moderatorRepository;
+
+  @Autowired
+  PasswordEncoder passwordEncoder;
+
+  /**
+   * This service method modifies an exisiting moderator based on the given inputs
+   * 
+   * Note we do not modify the role of that app user, because we consider it to
+   * stay a moderator.
+   * We can use the modifyUser in the appUser service to change its role
+   * 
+   * @param email       new email of the moderator (must be unique)
+   * @param lastname    new last name of the moderator
+   * @param firstname   new first name of the moderator
+   * @param authorities new authorities of the moderator
+   * @return the modified moderator
+   */
+  @Transactional
+  public AppUser modifyModerator(String email, String lastname, String firstname, Set<Authority> authorities) {
+
+    // Input Validation
+    String error = "";
+    if (email == null || email.isEmpty()) {
+      error += "Email cannot be left empty! ";
+    }
+    if (firstname == null || firstname.isEmpty()) {
+      error += "First name cannot be left empty! ";
+    }
+    if (lastname == null || lastname.isEmpty()) {
+      error += "Last name cannot be left empty! ";
+    }
+    if (authorities == null) {
+      error += "Authorities cannot be null! ";
+    }
+    if (error.length() > 0) {
+      throw new GlobalException(HttpStatus.BAD_REQUEST,
+          error);
+    }
+
+    // Check if the oderator we are trying to modify does indeed exist
+    AppUser moderator = appUserRepository.findAppUserByEmail(email);
+    if (moderator == null) {
+      throw new GlobalException(HttpStatus.BAD_REQUEST,
+          "This account does not exist.");
+    }
+
+    // Check that the new auhtorities of that user still include a moderator
+    Iterator<Authority> itr = authorities.iterator();
+    boolean isModerator = false;
+    while (itr.hasNext()) {
+      if (itr.next().equals(Authority.Moderator)) {
+        isModerator = true;
+      }
+    }
+    if (!isModerator) {
+      throw new GlobalException(HttpStatus.BAD_REQUEST,
+          "New authorities do not include the moderator authority!");
+    }
+
+    List<Role> roles = moderator.getRole();
+
+    boolean modified = false;
+
+    for (int i = 0; i < roles.size(); i++) {
+      if (roles.get(i) instanceof Moderator) {
+        moderator.setFirstname(firstname);
+        moderator.setLastname(lastname);
+        moderator.setEmail(email);
+        moderator.setAuthorities(authorities);
+        appUserRepository.save(moderator);
+        modified = true;
+      }
+    }
+    if (!modified) {
+      throw new GlobalException(HttpStatus.BAD_REQUEST,
+          "User is not a moderator!");
+    }
+    return moderator;
+  }
+
+  /**
+   * This service method updates the moderator's password based on the givenm
```

Comment:
There seems to be a similar function in AppUserService
Is this function really necessary? This one restricts the modification to the moderator but the other one doesn't restrict the role
It looks like the javadoc from the function in the AppUserService is copy pasted from here and doesn't completely correspond